src/sql/src/pure.rs

```diff
@@ -560,6 +511,123 @@ async fn purify_csr_connector(
+                    .has_headers(false)
+                    .from_reader(csv_header.as_bytes());
+
+                if let Some(result) = reader.records().next() {
+                    match result {
+                        Ok(headers) => Some(headers),
+                        Err(e) => bail!("Unable to parse header row: {}", e),
+                    }
+                } else {
+                    None
+                }
+            }
+            Ok(None) => {
+                if let CsvColumns::Header { names } = columns {
+                    if names.is_empty() {
+                        bail!(
+                            "CSV file expected to have at least one line \
+                                     to determine column names, but is empty"
+                        );
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+            Err(e) => {
+                // TODO(#7562): support compressed files
+                if let CsvColumns::Header { names } = columns {
+                    if names.is_empty() {
+                        bail!("Cannot determine header by reading CSV file: {}", e);
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+        }
+    } else {
+        None
+    };
+    match (&columns, first_row) {
+        (CsvColumns::Header { names }, Some(headers)) if names.is_empty() => {
+            *columns = CsvColumns::Header {
+                names: headers.into_iter().map(Ident::from).collect(),
+            };
+        }
+        (CsvColumns::Header { names }, Some(headers)) => {
+            if names.len() != headers.len() {
+                bail!(
+                    "Named column count ({}) does not match \
+                               number of columns discovered ({})",
+                    names.len(),
+                    headers.len()
+                );
+            } else if let Some((sql, csv)) = names
+                .iter()
+                .zip(headers.iter())
+                .find(|(sql, csv)| sql.as_str() != &**csv)
+            {
+                bail!(
+                    "Header columns do not match named columns from CREATE SOURCE statement. \
+                               First mismatched columns: {} != {}",
+                    sql,
+                    csv
+                );
+            }
+        }
+        (CsvColumns::Header { names }, None) if names.is_empty() => {
+            bail!(
+                "WITH HEADER requires a way to determine the header row, but file does not exist"
+            );
+        }
+        (CsvColumns::Header { names }, None) => {
+            // we don't need to do any verification if we are told the names of the headers
+            assert!(!names.is_empty(), "match arm moved into the wrong spot");
```

Comment:
"match arm moved into the wrong spot" is a bit unclear to me

Reply:
Yeah unfortunately now this match statement is sort of order dependent, although let me check if we fail CI for dead code, which would make this unnecessary.

Reply:
yeah this assert is useless we do fail on dead code.